ISPN-6576: backport to 8.2.x

ksobolew opened the pull request:
The tests are a little different than the ones attached to the issue - there are fewer classes and the tests in transactional mode are disabled (because of ISPN-6573). The fix is almost too simple to make me confident it's the right one :)

rvansa commented:
Do you think it would be possible to just add the test as cases to the existing tests, rather than adding a completely new ones? And you can drop the tx-support in the tests, that can be added later when it's implemented.

rvansa commented:
(and btw., when updating PR, you don't have to close the old one and create new, just do a force-push into your branch).

ksobolew commented:
Yeah, I closed the PR accidentally when I was trying to put these commits on a separate branch :)

As for merging these tests with some existing ones - well, I don't know. They need some specific configuration, but I'll see what I can do. (Oh, and I can't seem to get TestNG to work right, *sigh*)

ksobolew commented:
I don't think I can find another home for these tests. I removed the transactional mode tests and simplified the rest a little (at the cost of parameterization of AbstractFunctionalTest). And I fixed them so that they actually work :)

ksobolew commented:
Ah, one more thing: the tests will pass only when ISPN-6542 is fixed; until then they'll throw a NPE. I forgot that I had the proposed fix for that committed on my local branch.

ksobolew commented:
Another bunch of tests and fixes. In addition to the simple ReadWriteMap.eval() the tests now also exercise all other types of functional Maps and operations on them. Note that it still requires the fix to ISPN-6542; see also https://github.com/infinispan/infinispan/pull/4265#issuecomment-217165678 which I originally did for this branch but it depends on rvansa's commit there.

ksobolew commented:
Argh, I see that the ManyDataCommand interface was also added by rvansa for ISPN-6542 and it's missing here, so it does not compile. Oh well...

ksobolew commented:
Rebased on top of back-ported ISPN-6542.

galderz commented:
Hey @ksobolew, thanks a lot for sending the PR. I'll have a look at the PR tomorrow and get back to you. Thanks for getting involved with the project 👏 

@rvansa Thanks for the help with this 😃

ksobolew commented:
@galderz You're walcome :) I'm sort of familiar with the codebase from the JBoss Cache days (we're finally migrating to Infinispan!) so I could just dive in and try to fix the problem myself

galderz commented on commons/src/main/java/org/infinispan/commons/api/functional/Param.java:
```diff
@@ -44,6 +44,29 @@
+   /**
+    * The way how this object should be serialized;
+    */
+   enum SerializeMethod {
```
This is not how we deal with an objects serialization/marshalling characteristics. Instead, if a param instance needs to be serialized, an Externalizer should be provided for it (examples in the codebase) and then the Externalizer it's either caught by annotation processing (e.g. @SerializedWith) or it's preregistered. The externalizer itself can decide how it serialize it using whichever method it wants. A param implementation, if you want something lazy, can implement Serializable and that's it. So, there's no need to add anything to the Param interface, which has been designed with the bare minimum required to do its job.

galderz replied:
Also, whether the param itself should be sent over the wire or not is not down the parameter itself, but more to the command being executed. The command itself must decide whether it includes certain params inside its externalizer payload. Params themselves do not live on their own.

ksobolew replied:
Note that this is part of back-ported #4265 by @rvansa

galderz replied:
Cross posted the comments to that PR.

galderz commented on core/src/main/java/org/infinispan/commands/functional/ManyDataCommand.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public interface ManyDataCommand<K> {
```
I don't really see the need to add yet another interface to the huge chain of *Command of interfaces we already have. What I mean by that is this: Each AbstractWriteManyCommand implementation could have a getKeys() method inside, that does not depend on an interface, and when it arrives to the interceptor level, each instance of the command, for which we have a visit* call, could call its own getKeys() and pass that to the method.

In fact, I think adding this interface makes the code more complex unnecessarily by having to use `&` structures to find commands that are X and Y...etc.

Truth to be said, I did the same error with ParamsCommand, and I can see now how this could get out for unnecessary reasons really.

rvansa replied:
That's a valid point. I'll update #4265 accordingly.

danberindei replied:
@galderz I don't really agree this makes the code more complex. There's a lot of duplicate code, e.g. in `NonTxDistributionInterceptor`, that only exists because we don't have an interface with `getKeys()` on it.

galderz replied:
I agree there's duplicate code and to be honest that class could do with some refactoring to avoid that. But I don't know for sure if your point is about duplication between different functional command visit impls or between functional command visit impls and the others.

In any case, I'm happy for such ManyDataCommand interface to exist if it's really going to reduce some code in that class too.

galderz commented on core/src/main/java/org/infinispan/functional/impl/Params.java:
```diff
@@ -108,4 +115,50 @@ public static Params from(Param<?>... ps) {
+      @Override
+      public void writeObject(ObjectOutput output, Params object) throws IOException {
+         for (Param param : object.params) {
+            switch (param.serializeMethod()) {
```
I see what you are doing here but this is brittle IMO. Simply have each Param have its own Externalizer and that should work without any further complications nor the need to add an extra method to the interface. Besides, I don't expect that many params to make this is really that useful. IOW, I expect the number of Param impls to be in the tens and not in the hundreds :)

rvansa replied:
@galderz So you think that Params should already know about all possible Param implementations (it does, through defaults), and serialize them one-by-one? That's what I've tried to avoid, by delegating the information about the params to themselves.
I want to spare the bytes and lookups in marshaller when the params always have the same order (according to their ids) and some of them don't make sense to be serialized at all (such as the future mode which is purely local thing).

galderz replied:
@rvansa Param implementations will be provided by us, it's like flags really. Params control how the internals work, e.g. whether to check the cache store or not...etc, so really they are something we control. I don't expect users to implement Param instances. On the other hand, MetaParam instances can easily be provided by the user since they're stored along with the entry (like Metadata) so we have less control there.

I can see the utility of having a single externalizer for all Params, but without your extra method you'd require some kind of lookup. Besides, I don't think your method would work for Param instances that carry values.

One externalizer per Param would just work fine IMO.

rvansa replied:
MetaParams are different thing; I haven't touched those. I understand why these need externalizer per MetaParam.

But as for Params, we control those, so we can handle them in single Params.Externalizer. I'll push update to the other PR, so please let's move the discussion there.

galderz commented on core/src/main/java/org/infinispan/interceptors/ClusteredCacheLoaderInterceptor.java:
```diff
@@ -41,24 +41,35 @@ private void startClusteredCacheLoaderInterceptor() {
 
    @Override
-   protected boolean skipLoadForWriteCommand(WriteCommand cmd, Object key, InvocationContext ctx) {
+   protected boolean skipLoadForWriteCommand(WriteCommand cmd, Object key, InvocationContext ctx, boolean functionalReadWrite) {
```
I'm not sure I like this boolean parameter either. It makes the method quite complex adding further if statements and it's already quite difficult to read. I think i'd rather add some code duplication by having functional-specific methods which are more focused, smaller, and easier to read.

Same thing for other methods that have been extended to have `functionalReadWrite` boolean parameters

ksobolew replied:
Right. I was not sure about that either. Earlier versions (still immortalized by earlier commits in this PR) used instanceof checks deeper in the call stack but I was even less sure about those :) Would these checks be more acceptable or should I go the functional-specific methods way?

galderz replied:
I think I'd go for functional-specific methods.

galderz commented:
@ksobolew Radim's changes are in, so make sure you rebase with latest changes he made :)

ksobolew commented:
@galderz I'm on it!

ksobolew commented:
Updated and rebased.

ksobolew commented on core/src/main/java/org/infinispan/interceptors/CacheLoaderInterceptor.java:
```diff
@@ -415,7 +444,12 @@ private boolean skipLoad(FlagAffectedCommand cmd, Object key, InvocationContext
 
       boolean skip;
-      if (cmd instanceof WriteCommand) {
+      if (cmd instanceof AbstractWriteKeyCommand || cmd instanceof AbstractWriteManyCommand) {
```
Since the line 452 (formerly 418) uses instanceof, I thought I might as well add some more tests like this. I'm not etirely confident that testing for Abstract* classes is the way to go - maybe I should add a marker interface FunctionalCommand?

rvansa commented:
Sorry @ksobolew, @danberindei pointed out some shortcomings so I've added one more commit as #4359

ksobolew commented:
Any progress on this?

galderz commented:
@ksobolew Sorry for the delay, I'll have a look into this later on today.

galderz commented:
Why is this PR against 8.2.x? We're only fixing critical stuff in 8.x. This fix should be sent first for master and then we can backport if necessary. I'll do a quick review in the mean time but it's confusing since ISPN-6542 is already fixed in master and don't want to go over the same code again. Plus, the fact other fixes are quite spread (would it make sense to squash them?) does not make review easy :|.

galderz commented on core/src/test/java/org/infinispan/functional/FunctionalDistributionTest.java:
```diff
@@ -0,0 +1,108 @@
+            }).join();
+
+      // since we're in async mode, we need to wait before the replication finishes
+      Thread.sleep(100); // TODO: find a better way?
```
Minor thing, TestingUtil.sleepThread() sleeps and does not throw InterruptedException.

galderz commented:
Looks good otherwise, thanks a lot @ksobolew 😃. Please send a fix for master and try, if possible, to squash the commits if it makes sense to do so.

ksobolew commented:
@galderz Yes, sorry about the mixup, I should've sent a PR with the 6542 backport first (it's #4357 , BTW), then base this PR on that. But you're right, I also should've familiarized myself more with the procedures :) I'm working with 8.2.x on my application so that's where I fix the problems I encounter. I'll rebase it on top of master and push again soon, thanks!

ksobolew commented:
Rebased as a backport of #4378 to 8.2.x

@galderz I hope that this falls under "backport if necessary" :) This is pretty critical for me because the Functional API does not work at all and I'm relying on it.

galderz commented:
Integrated, thanks a million @ksobolew 😃
